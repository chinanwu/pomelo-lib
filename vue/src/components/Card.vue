<template>
  <div
    :id="id"
    class="Card"
    :class="[{ 'Card--flip': isFlipped }, cClass]"
    tabindex="0"
    @click="handleClick"
    @keydown="handleKeyDown"
  >
    <div class="Card__front">
      <slot name="front"></slot>
    </div>
    <div class="Card__back">
      <slot name="back"></slot>
    </div>
  </div>
</template>

<script>
export default {
  name: "Card",
  props: {
    id: String,
    cClass: String,
    flipped: {
      type: Boolean,
      default: false
    }
  },
  data() {
    return {
      isFlipped: this.flipped
    };
  },
  emits: ["click"],
  methods: {
    handleClick() {
      this.isFlipped = !this.isFlipped;
      this.$emit("click");
    },
    handleKeyDown(event) {
      if (
        event &&
        event.keyCode &&
        !event.shiftKey &&
        !event.ctrlKey &&
        !event.altKey &&
        !event.metaKey &&
        (event.keyCode === 13 || event.keyCode === 32) // if enter or space is pressed
      ) {
        event.preventDefault();
        this.handleClick();
      }
    }
  }
};
</script>

<style lang="less" scoped>
.Card {
  width: 40rem;
  height: 20rem;
  transform-origin: 50% 50%;
  transform-style: preserve-3d;
  transition: transform 0.25s ease-in-out, box-shadow 0.3s;
  cursor: pointer;
}

.Card--flip {
  transform: rotateY(180deg);
}

.Card__front,
.Card__back {
  display: flex;
  align-items: center;
  position: absolute;
  top: 0;
  left: 0;
  backface-visibility: hidden;
  width: 100%;
  height: 100%;
  background-color: white;
}

.Card__back {
  transform: rotateY(180deg);
}
</style>
