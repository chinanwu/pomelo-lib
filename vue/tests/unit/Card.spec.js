import { mount } from "@vue/test-utils";
import Card from "@/components/Card.vue";

describe("Card.vue", () => {
  it("renders card", () => {
    const wrapper = mount(Card);
    expect(wrapper.element).toMatchSnapshot();
  });

  it("renders card with front and back slots", () => {
    const wrapper = mount(Card, {
      slots: {
        front: "<div class='test'>front</div>",
        back: "<div class='test'>front</div>"
      }
    });
    expect(wrapper.findAll(".test").length).toBe(2);
  });

  it("flips when clicked", () => {
    const wrapper = mount(Card, {
      props: {
        id: "test"
      }
    });
    expect(wrapper.element).toMatchSnapshot();
    wrapper.find("#test").trigger("click");
    expect(wrapper.emitted()).toHaveProperty("click");
  });
});
