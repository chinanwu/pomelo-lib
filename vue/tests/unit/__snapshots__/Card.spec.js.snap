// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.vue flips when clicked 1`] = `
<div
  class="Card"
  id="test"
  tabindex="0"
>
  <div
    class="Card__front"
  >
    
    
  </div>
  <div
    class="Card__back"
  >
    
    
  </div>
</div>
`;

exports[`Card.vue flips when the enter or space button is pressed 1`] = `
<div
  class="Card"
  id="test"
  tabindex="0"
>
  <div
    class="Card__front"
  >
    
    
  </div>
  <div
    class="Card__back"
  >
    
    
  </div>
</div>
`;

exports[`Card.vue renders card 1`] = `
<div
  class="Card"
  tabindex="0"
>
  <div
    class="Card__front"
  >
    
    
  </div>
  <div
    class="Card__back"
  >
    
    
  </div>
</div>
`;
